### DISCUSSION TITLE
general troubleshooting connectivity problems (SPI)

### DISCUSSION THREAD

**Discussion by jschoch** [multigcs/LinuxCNC-RIO#30] [General]
Ok, i made a new board and setup for SPI0 and got it out of e-stop and ran the test gcode.  Then I hooked up my scope, saw a brief set of stepper pulses and then poof it stopped working.  I can connect via the gowin programmer, i reprogrammed it, but qt_spitest just shows 0 repose bytes and I'm at a loss as to how to do anything else. 

Is there any way to ensure the board is alive?  The blink on pin 10, is that the blinking orange LED on the tang nano 9k?  I'm a complete noob in FPGA land and I have not the clue to figurate a first place to test/troubleshoot this board.  I'm not even sure how to reset it beyond plug/unplug.

Is there any way to get serial or status from the 9k?

since the loader doesn't work on the pi, i'm not sure if I should run stuff on the pi or on my PC.  I had to SCP over the Output folders.  is this the right way to do it?

If i want to reduce the SPI speed for trobleshooting do I just do that in the qt_spitest app, or do I need to update the 9k build and reflash?

is there any documentation of the led's/status codes/blink codes?

is the default spi=on ok for the /boot/config.txt?  It seems this uses CS/CE1 but when I run

```
schoch@remoraPi4:~ $ sudo raspi-gpio get
BANK0 (GPIO 0 to 27):
GPIO 0: level=1 fsel=0 func=INPUT pull=UP
GPIO 1: level=1 fsel=0 func=INPUT pull=UP
GPIO 2: level=1 fsel=0 func=INPUT pull=UP
GPIO 3: level=1 fsel=0 func=INPUT pull=UP
GPIO 4: level=1 fsel=0 func=INPUT pull=UP
GPIO 5: level=1 fsel=0 func=INPUT pull=UP
GPIO 6: level=1 fsel=0 func=INPUT pull=UP
GPIO 7: level=1 fsel=1 func=OUTPUT pull=UP
GPIO 8: level=1 fsel=1 func=OUTPUT pull=UP
GPIO 9: level=0 fsel=4 alt=0 func=SPI0_MISO pull=DOWN
GPIO 10: level=1 fsel=4 alt=0 func=SPI0_MOSI pull=DOWN
GPIO 11: level=0 fsel=4 alt=0 func=SPI0_SCLK pull=DOWN
GPIO 12: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 13: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 14: level=1 fsel=4 alt=0 func=TXD0 pull=NONE
GPIO 15: level=1 fsel=4 alt=0 func=RXD0 pull=UP
GPIO 16: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 17: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 18: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 19: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 20: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 21: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 22: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 23: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 24: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 25: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 26: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 27: level=0 fsel=0 func=INPUT pull=DOWN
BANK1 (GPIO 28 to 45):
GPIO 28: level=1 fsel=2 alt=5 func=RGMII_MDIO pull=UP
GPIO 29: level=0 fsel=2 alt=5 func=RGMII_MDC pull=DOWN
GPIO 30: level=1 fsel=0 func=INPUT pull=UP
GPIO 31: level=1 fsel=0 func=INPUT pull=NONE
GPIO 32: level=1 fsel=0 func=INPUT pull=DOWN
GPIO 33: level=1 fsel=0 func=INPUT pull=DOWN
GPIO 34: level=0 fsel=7 alt=3 func=SD1_CLK pull=NONE
GPIO 35: level=1 fsel=7 alt=3 func=SD1_CMD pull=UP
GPIO 36: level=1 fsel=7 alt=3 func=SD1_DAT0 pull=UP
GPIO 37: level=1 fsel=7 alt=3 func=SD1_DAT1 pull=UP
GPIO 38: level=1 fsel=7 alt=3 func=SD1_DAT2 pull=UP
GPIO 39: level=1 fsel=7 alt=3 func=SD1_DAT3 pull=UP
GPIO 40: level=0 fsel=4 alt=0 func=PWM1_0 pull=NONE
GPIO 41: level=0 fsel=4 alt=0 func=PWM1_1 pull=NONE
GPIO 42: level=0 fsel=1 func=OUTPUT pull=UP
GPIO 43: level=1 fsel=0 func=INPUT pull=UP
GPIO 44: level=1 fsel=0 func=INPUT pull=UP
GPIO 45: level=1 fsel=0 func=INPUT pull=UP
BANK2 (GPIO 46 to 53):
GPIO 46: level=0 fsel=0 func=INPUT pull=UP
GPIO 47: level=0 fsel=0 func=INPUT pull=UP
GPIO 48: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 49: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 50: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 51: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 52: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 53: level=0 fsel=0 func=INPUT pull=DOWN

``` 

I don't see the CE pins, based on a previous comment and the rio.c halcomponent it looks like you manually control the CS pin directly, not sure the implications of the different pin mappings for PI 2,3,4,5 since it is using a different constant referencing the chip pin and not the coloquial GPIO numbering.

testing via spi_test shows;

```
sudo /home/alex/spi/spidev-test/spidev_test -v -D /dev/spidev0.1
spi mode: 0x4
bits per word: 8
max speed: 500000 Hz (500 KHz)
TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  | ......@....▒..................▒.
RX | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  | ................................
```

Same for /dev/spidev0.0

Is spidev ok or should it be disabled?

**Comment by multigcs** [comment]
You can use scp or better rsync to copy your Output-Folder to the raspberry

You do not need to change any speed settings, all should works as it is

the blinking LED is to see if the FPGA works (1Hz fixed).

please use the python tool to test the SPI connection:

`python3 Output/TangNano9K/Firmware/qt_spitest.py`

there you can also see the SPI response.

**Reply by jschoch** [reply]
Oddly when i try to run linux cnc the CE0 pin gets changed

```
GPIO 8: level=1 fsel=4 alt=0 func=SPI0_CE0_N pull=UP
GPIO 9: level=0 fsel=4 alt=0 func=SPI0_MISO pull=DOWN
GPIO 10: level=1 fsel=4 alt=0 func=SPI0_MOSI pull=DOWN
GPIO 11: level=0 fsel=4 alt=0 func=SPI0_SCLK pull=DOWN

```

Also wondering if the linuxcnc version i am running may be a problem, it is LINUXCNC - 2.10.0~pre0

re the test program it doesn't see anything and while I can see signals on MOSI nothing happens on the MISO pin.

the output is 

```
tx: [116, 105, 114, 119, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 0, 0]
Duration: 2.88ms
rx: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
ERROR: Unknown Header: 0x0

tx: [116, 105, 114, 119, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 0, 0]
Duration: 2.90ms
rx: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
ERROR: Unknown Header: 0x0

tx: [116, 105, 114, 119, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 0, 0]
Duration: 2.88ms
rx: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
ERROR: Unknown Header: 0x0

... forever

```

again, i'm a bit concerned i did something to the chip when I hooked the probe up to it since it was working and now nothing.

I changed the clock in the config.json (deleted a zero) and now the LED is blinking much faster (a bit counter intuitive)...

about to give up and smash all the technology in my house lol, only insane people enjoy this stuff (that is self criticism).

**Comment by jschoch** [comment]
I just noticed that the rio.c in my Output folder has the pin for CS/CE0 set (RPI_GPIO_P1_24) and not CS1 (RPI_GPIO_P1_26) per the BOB board and the readme/install document for the tang 9k.  the comment also says //CS0  Is this an error with the config or template?

**Comment by multigcs** [comment]
if you remove a zero in the clock value, than the LED runs faster (10Hz) because the real crystal keeps the same speed,
only the divider changes.

the right CS connection is:

`CS (Pin 32 on FPGA) -> GPIO7 (CE1 on Raspberry)`

the rio.c looks a little confusing, i have to clean it

**Comment by jschoch** [comment]
can i setup a blink on one of the 3.3v pins to see if they are working?  My scope is a DS1054, not sure I can see super fast signals.  what else can I probe to see if it is alive?

**Comment by multigcs** [comment]
whats wrong with the blinking LED ?
it's 1Hz, no need for a scope :)

**Reply by jschoch** [reply]
my fear is i burnt out the 3.3v pins somehow.

**Reply by multigcs** [reply]
replace the rio.v with this:
```

module rio (
        output BLINK_LED,
        output DIN0,
        output DIN1,
        output DIN2,
        output DIN3,
        output DIN4,
        output DIN5,
        output DIN6,
        output DIN7,
        output DIN8,
        output DIN9,
        output DIN10,
        output DIN11,
        output DIN12,
        output DOUT0,
        output DOUT1,
        output DOUT2,
        output DOUT3,
        output ENA,
        output ERROR_OUT,
        output EXPANSION0_SHIFTREG_CLOCK,
        output EXPANSION0_SHIFTREG_LOAD,
        output EXPANSION0_SHIFTREG_OUT,
        output EXPANSION0_SHIFTREG_IN,
        output INTERFACE_SPI_MOSI,
        output INTERFACE_SPI_MISO,
        output INTERFACE_SPI_SCK,
        output INTERFACE_SPI_SSEL,
        output JOINT35_STEPPER_STP,
        output JOINT35_STEPPER_DIR,
        output JOINT36_STEPPER_STP,
        output JOINT36_STEPPER_DIR,
        output JOINT37_STEPPER_STP,
        output JOINT37_STEPPER_DIR,
        output JOINT38_STEPPER_STP,
        output JOINT38_STEPPER_DIR,
        output JOINT39_STEPPER_STP,
        output JOINT39_STEPPER_DIR,
        output VIN33_PULSECOUNTER_UP,
        output VIN33_PULSECOUNTER_DOWN,
        output VOUT34_PWM_PWM,
        input sysclk
    );


    reg ESTOP = 0;
    wire ERROR;
    wire INTERFACE_TIMEOUT;
    assign ERROR = (INTERFACE_TIMEOUT | ESTOP);
    blink #(135000) blink1 (
        .clk (sysclk),
        .led (BLINK_LED)
    );

    assign DIN0 = BLINK_LED;
    assign DIN1 = BLINK_LED;
    assign DIN2 = BLINK_LED;
    assign DIN3 = BLINK_LED;
    assign DIN4 = BLINK_LED;
    assign DIN5 = BLINK_LED;
    assign DIN6 = BLINK_LED;
    assign DIN7 = BLINK_LED;
    assign DIN8 = BLINK_LED;
    assign DIN9 = BLINK_LED;
    assign DIN10 = BLINK_LED;
    assign DIN11 = BLINK_LED;
    assign DIN12 = BLINK_LED;
    assign DOUT0 = BLINK_LED;
    assign DOUT1 = BLINK_LED;
    assign DOUT2 = BLINK_LED;
    assign DOUT3 = BLINK_LED;
    assign ENA = BLINK_LED;
    assign ERROR_OUT = BLINK_LED;
    assign EXPANSION0_SHIFTREG_CLOCK = BLINK_LED;
    assign EXPANSION0_SHIFTREG_LOAD = BLINK_LED;
    assign EXPANSION0_SHIFTREG_OUT = BLINK_LED;
    assign EXPANSION0_SHIFTREG_IN = BLINK_LED;
    assign INTERFACE_SPI_MOSI = BLINK_LED;
    assign INTERFACE_SPI_MISO = BLINK_LED;
    assign INTERFACE_SPI_SCK = BLINK_LED;
    assign INTERFACE_SPI_SSEL = BLINK_LED;
    assign JOINT35_STEPPER_STP = BLINK_LED;
    assign JOINT35_STEPPER_DIR = BLINK_LED;
    assign JOINT36_STEPPER_STP = BLINK_LED;
    assign JOINT36_STEPPER_DIR = BLINK_LED;
    assign JOINT37_STEPPER_STP = BLINK_LED;
    assign JOINT37_STEPPER_DIR = BLINK_LED;
    assign JOINT38_STEPPER_STP = BLINK_LED;
    assign JOINT38_STEPPER_DIR = BLINK_LED;
    assign JOINT39_STEPPER_STP = BLINK_LED;
    assign JOINT39_STEPPER_DIR = BLINK_LED;
    assign VIN33_PULSECOUNTER_UP = BLINK_LED;
    assign VIN33_PULSECOUNTER_DOWN = BLINK_LED;
    assign VOUT34_PWM_PWM = BLINK_LED;

endmodule

```
i can not test it yet, but it should output a 100Hz signal on most pins

**Comment by jschoch** [comment]
This was super helpful, it helped eliminate the 9k as the issue.  I updated/loaded this and it was blinking everything great!

Then, despite me checking my adapter board at least 15 times, I pulled it out and slapped in the crappy dupont wires.  Snap, worked great so it was my diy adapter board...  Need to get it under the scope to see WTF is wrong with it and why continuity testing didn't identify issues.

I want to see how it behaves with the CE/CS0 pin set back to the original value.  this may have been a combination of my home made pcb and that pin in the rio.c halcomponent.

This is likely a pretty good troubleshooting step.

**Reply by jschoch** [reply]
oddly the pin setting RPI_GPIO_P1_24 or RPI_GPIO_P1_26 in rio.c halcomponent doesn't seem to matter which doesn't make any sense.

**Comment by jschoch** [comment]
Another oddity is that after i ran the first test of the linuxcnc logo gcode the x step pin was kinda flopping on/off every 5-6 seconds.  it stopped after i hit the machine power gui btn.... Very odd.

**Reply by jschoch** [reply]
ok, it is doing it again... not sure what to help you out with but this seems odd.   it is the standard tango 9k bob pin 51, pin 54 is also doing it.  it is not a regular interval and maybe noise from the dupont cables?

**Reply by multigcs** [reply]
the 1 step back and forward is a known issue, it's a problem with the PID controller and i have no solution for this.
